File: CyxbsMobile2019_iOS/Other/DeviceUtils.m
//
//  DeviceUtils.m
//  CyxbsMobile2019_iOS
//
//  Created by Stove on 2021/5/26.
//  Copyright © 2021 Redrock. All rights reserved.
//

#import "DeviceUtils.h"
#import <sys/utsname.h>
static DeviceUtils* __deviceUtilsOne;
static DeviceType DeviceUtilsType;

@implementation DeviceUtils
+(void)load {
    DeviceUtilsType = 0;
    __deviceUtilsOne = [[super alloc] init];
}

+ (instancetype)alloc {
    if (__deviceUtilsOne==nil) {
        __deviceUtilsOne = [super alloc];
    }
    return __deviceUtilsOne;
}

+ (DeviceType)deviceType {
    return DeviceUtilsType;
}

DeviceType getDeviceType(void){
    
    struct utsname systemInfo;
    uname(&systemInfo);
    NSString *platform = [NSString stringWithCString:systemInfo.machine
                                           encoding:NSUTF8StringEncoding];
    
    //simulator
    if ([platform isEqualToString:@"i386"])          return Simulator;
    if ([platform isEqualToString:@"x86_64"])        return Simulator;

    //iPhone
    if ([platform isEqualToString:@"iPhone1,1"])     return IPhone_1G;
    if ([platform isEqualToString:@"iPhone1,2"])     return IPhone_3G;
    if ([platform isEqualToString:@"iPhone2,1"])     return IPhone_3GS;
    if ([platform isEqualToString:@"iPhone3,1"])     return IPhone_4;
    if ([platform isEqualToString:@"iPhone3,2"])     return IPhone_4;
    if ([platform isEqualToString:@"iPhone4,1"])     return IPhone_4s;
    if ([platform isEqualToString:@"iPhone5,1"])     return IPhone_5;
    if ([platform isEqualToString:@"iPhone5,2"])     return IPhone_5;
    if ([platform isEqualToString:@"iPhone5,3"])     return IPhone_5C;
    if ([platform isEqualToString:@"iPhone5,4"])     return IPhone_5C;
    if ([platform isEqualToString:@"iPhone6,1"])     return IPhone_5S;
    if ([platform isEqualToString:@"iPhone6,2"])     return IPhone_5S;
    if ([platform isEqualToString:@"iPhone7,1"])     return IPhone_6P;
    if ([platform isEqualToString:@"iPhone7,2"])     return IPhone_6;
    if ([platform isEqualToString:@"iPhone8,1"])     return IPhone_6s;
    if ([platform isEqualToString:@"iPhone8,2"])     return IPhone_6s_P;
    if ([platform isEqualToString:@"iPhone8,4"])     return IPhone_SE;
    if ([platform isEqualToString:@"iPhone9,1"])     return IPhone_7;
    if ([platform isEqualToString:@"iPhone9,3"])     return IPhone_7;
    if ([platform isEqualToString:@"iPhone9,2"])     return IPhone_7P;
    if ([platform isEqualToString:@"iPhone9,4"])     return IPhone_7P;
    if ([platform isEqualToString:@"iPhone10,1"])    return IPhone_8;
    if ([platform isEqualToString:@"iPhone10,4"])    return IPhone_8;
    if ([platform isEqualToString:@"iPhone10,2"])    return IPhone_8P;
    if ([platform isEqualToString:@"iPhone10,5"])    return IPhone_8P;
    if ([platform isEqualToString:@"iPhone10,3"])    return IPhone_X;
    if ([platform isEqualToString:@"iPhone10,6"])    return IPhone_X;
    if ([platform isEqualToString:@"iPhone11,8"])    return iPhone_XR;
    if ([platform isEqualToString:@"iPhone11,2"])    return iPhone_XS;
    if ([platform isEqualToString:@"iPhone11,4"] || [platform isEqualToString:@"iPhone11,6"])  return iPhone_XS_Max;

    return Unknown;
   
}
@end

/*
 + (DeviceType)deviceType{
     struct utsname systemInfo;
     uname(&systemInfo);
     NSString *platform = [NSString stringWithCString:systemInfo.machine
                                            encoding:NSUTF8StringEncoding];
     
     //simulator
     if ([platform isEqualToString:@"i386"])          return Simulator;
     if ([platform isEqualToString:@"x86_64"])        return Simulator;

     //iPhone
     if ([platform isEqualToString:@"iPhone1,1"])     return IPhone_1G;
     if ([platform isEqualToString:@"iPhone1,2"])     return IPhone_3G;
     if ([platform isEqualToString:@"iPhone2,1"])     return IPhone_3GS;
     if ([platform isEqualToString:@"iPhone3,1"])     return IPhone_4;
     if ([platform isEqualToString:@"iPhone3,2"])     return IPhone_4;
     if ([platform isEqualToString:@"iPhone4,1"])     return IPhone_4s;
     if ([platform isEqualToString:@"iPhone5,1"])     return IPhone_5;
     if ([platform isEqualToString:@"iPhone5,2"])     return IPhone_5;
     if ([platform isEqualToString:@"iPhone5,3"])     return IPhone_5C;
     if ([platform isEqualToString:@"iPhone5,4"])     return IPhone_5C;
     if ([platform isEqualToString:@"iPhone6,1"])     return IPhone_5S;
     if ([platform isEqualToString:@"iPhone6,2"])     return IPhone_5S;
     if ([platform isEqualToString:@"iPhone7,1"])     return IPhone_6P;
     if ([platform isEqualToString:@"iPhone7,2"])     return IPhone_6;
     if ([platform isEqualToString:@"iPhone8,1"])     return IPhone_6s;
     if ([platform isEqualToString:@"iPhone8,2"])     return IPhone_6s_P;
     if ([platform isEqualToString:@"iPhone8,4"])     return IPhone_SE;
     if ([platform isEqualToString:@"iPhone9,1"])     return IPhone_7;
     if ([platform isEqualToString:@"iPhone9,3"])     return IPhone_7;
     if ([platform isEqualToString:@"iPhone9,2"])     return IPhone_7P;
     if ([platform isEqualToString:@"iPhone9,4"])     return IPhone_7P;
     if ([platform isEqualToString:@"iPhone10,1"])    return IPhone_8;
     if ([platform isEqualToString:@"iPhone10,4"])    return IPhone_8;
     if ([platform isEqualToString:@"iPhone10,2"])    return IPhone_8P;
     if ([platform isEqualToString:@"iPhone10,5"])    return IPhone_8P;
     if ([platform isEqualToString:@"iPhone10,3"])    return IPhone_X;
     if ([platform isEqualToString:@"iPhone10,6"])    return IPhone_X;
     if ([platform isEqualToString:@"iPhone11,8"])    return iPhone_XR;
     if ([platform isEqualToString:@"iPhone11,2"])    return iPhone_XS;
     if ([platform isEqualToString:@"iPhone11,4"] || [platform isEqualToString:@"iPhone11,6"])  return iPhone_XS_Max;

     return Unknown;
    
 }
 */

File: CyxbsMobile2019_iOS/Tool/Other/DeviceUtils.swift
//
//  DeviceTest.swift
//  CyxbsMobile2019_iOS
//
//  Created by Stove on 2021/5/30.
//  Copyright © 2021 Redrock. All rights reserved.
//

//参考文章：https://www.jianshu.com/p/d920885d6a02

import UIKit

@objc enum DeviceType:Int {
    case Unknown = 0,
    Simulator,
    IPhone_1G,          //基本不用
    IPhone_3G,          //基本不用
    IPhone_3GS,         //基本不用
    IPhone_4,           //基本不用
    IPhone_4s,          //基本不用
    IPhone_5,
    IPhone_5C,
    IPhone_5S,
    IPhone_SE1,
    IPhone_6,
    IPhone_6P,
    IPhone_6s,
    IPhone_6s_P,
    IPhone_7,
    IPhone_7P,
    IPhone_8,
    IPhone_8P,
    IPhone_X,
    iPhone_XR,
    iPhone_XS,
    iPhone_XS_Max
}

@objc enum DeviceHardwareLevel:Int {
    case Low, Medium, High
}

public class DeviceUtil: NSObject {
    
    private var _type:DeviceType = .Unknown
    /// 如果是模拟器，那么返回Simulator
    @objc var type:DeviceType {
        get {
            return _type
        }
    }
    
    private var _hardwareLevel:DeviceHardwareLevel = .Medium
    /// 如果是模拟器，那么返回Low
    @objc var hardwareLevel:DeviceHardwareLevel {
        get {
            return _hardwareLevel
        }
    }
    
    
    public override init() {
        super.init()
        _type = getType()
        _hardwareLevel = getHardwareLevel()
    }
    
    private func getHardwareLevel() -> DeviceHardwareLevel {
        switch type.rawValue {
        case ..<DeviceType.IPhone_7.rawValue:
            return .Low
        case DeviceType.IPhone_7.rawValue..<DeviceType.IPhone_X.rawValue:
            return .Medium
        default:
            return .High
        }
    }
    
    private func getType() -> DeviceType {
        var systemInfo = utsname()
        uname(&systemInfo)
        //下面几行代码大概意思是把systemInfo.machine的数据转位String格式
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
           guard let value = element.value as? Int8, value != 0 else { return identifier }
           return identifier + String(UnicodeScalar(UInt8(value)))
        }
//        UIDevice.current.name;
        let t:DeviceType
        switch identifier {
        case "i386":
            t = .Simulator
        case "x86_64":
            t = .Simulator
        case "iPhone1,1":
            t = .IPhone_1G
        case "iPhone1,2":
            t = .IPhone_3G
        case "iPhone2,1":
            t = .IPhone_3GS
        case "iPhone3,1":
            t = .IPhone_4
        case "iPhone3,2":
            t = .IPhone_4
        case "iPhone4,1":
            t = .IPhone_4s
        case "iPhone5,1":
            t = .IPhone_5
        case "iPhone5,2":
            t = .IPhone_5
        case "iPhone5,3":
            t = .IPhone_5C
        case "iPhone5,4":
            t = .IPhone_5C
        case "iPhone6,1":
            t = .IPhone_5S
        case "iPhone6,2":
            t = .IPhone_5S
        case "iPhone7,1":
            t = .IPhone_6P
        case "iPhone7,2":
            t = .IPhone_6
        case "iPhone8,1":
            t = .IPhone_6s
        case "iPhone8,2":
            t = .IPhone_6s_P
        case "iPhone8,4":
            t = .IPhone_SE1
        case "iPhone9,1":
            t = .IPhone_7
        case "iPhone9,2":
            t = .IPhone_7P
        case "iPhone9,3":
            t = .IPhone_7
        case "iPhone9,4":
            t = .IPhone_7P
        case "iPhone10,1":
            t = .IPhone_8
        case "iPhone10,2":
            t = .IPhone_8P
        case "iPhone10,3":
            t = .IPhone_X
        case "iPhone10,4":
            t = .IPhone_8
        case "iPhone10,5":
            t = .IPhone_8P
        case "iPhone10,6":
            t = .IPhone_X
        case "iPhone11,2":
            t = .iPhone_XS
        case "iPhone11,4":
            t = .iPhone_XS_Max
        case "iPhone11,6":
            t = .iPhone_XS_Max
        case "iPhone11,8":
            t = .iPhone_XR
        default:
            t = .Unknown
    }
        return t
    }
}
